kythe/rust/indexer/src/indexer/analyzers.rs

```diff
@@ -719,6 +709,89 @@ impl<'a, 'b> CrateAnalyzer<'a, 'b> {
+            // If there is no reference id, we can't emit a cross reference
+            if reference.ref_id.is_none() {
+                continue;
+            }
+
+            // Currently this is only known to support `Use` ImportKinds
+            if reference.kind != rls_data::ImportKind::Use {
+                continue;
+            }
+
+            let ref_id = reference.ref_id.unwrap();
+
+            let mut reference_vname = template_vname.clone();
+            let span = &reference.span;
+
+            // Get the CrateId for the referenced crate
+            let krate_id_option = self.krate_ids.get(&ref_id.krate);
+            if krate_id_option.is_none() {
+                // This is a little bit of chicken and egg here. We could try
+                // and find the file vname first before the crate id, but then
+                // we wouldn't have a definition anchor to emit the diagnostic
+                // node to because we wouldn't have the crate id yet. The chance
+                // of a crate id not being in our HashMap is very low, so if this
+                // fails we just emit a diagnostic node to the first source file
+                // of the crate we are indexing
+                let file_vname = self.file_vnames.values().next().unwrap();
+                self.emitter.emit_diagnostic(
+                    file_vname,
+                    "Cross reference could not be generated",
+                    Some(&format!("Failed to generate cross reference for \"{:?}\" because the referenced crate could not be found", ref_id)),
+                    None
+                )?;
+            }
+            let krate_id = krate_id_option.unwrap();
+
+            // Create VName for target of reference
+            let mut target_vname = template_vname.clone();
+            target_vname.set_signature(self.create_def_signature(krate_id, ref_id.index));
+            target_vname.set_language("rust".to_string());
+
+            // Create VName for the reference node
+            let file_name = span.file_name.to_str().unwrap();
+            let file_vname = self.file_vnames.get(file_name);
+            if file_vname.is_none() {
+                self.emitter.emit_diagnostic(
+                    &target_vname,
+                    "Failed to get file VName for reference",
+                    Some(format!("The Rust indexer was unable to locate the file VName for the reference in the file \"{}\"", file_name).as_ref()),
+                    None,
+                )?;
+                continue;
+            }
+            reference_vname.set_path(file_vname.unwrap().get_path().to_string());
+
+            let byte_span_option = self.get_byte_span(&ref_id, span)?;
+            if byte_span_option.is_none() {
+                continue;
+            }
+            let byte_span = byte_span_option.unwrap();
+
+            // Create signature based on span
+            reference_vname.set_signature(format!(
+                "{}_ref_{}_{}",
```

Comment:
You do this twice; pull it out to a function that takes the krate_signature and the ByteSpan